### PR TITLE
Add molecular fingerprints and fingerprint based similarity

### LINKED ISSUE #42: Missing similarity measure based on molecular fingerprints.
In the spec2vec project we compare different spectral similarities to the similarity between respective __molecular fingerprints__ (as a kind of "ground truth"). 
Those fingerprints are calculated based on the actual structure of a molecule so they need smiles (or inchi) as input.  

There are many different flavors of molecular fingerprints. We implemented some key types using rdkit in the old spec2vec code. That should be refactored and added to matchms I think.

### PR DESCRIPTION
**Pull request by florian-huber**
Addressing #42 .

- Added ``add_fingerprint()`` filter that adds molecular fingerprint to spectrum object (if possible).
- Added ``FingerprintSimilarityParallel`` fo calculate cosine scores between molecular fingerprints. Since spectra without smiles or inchi will not have a fingerprint entry, the function returns either NaNs (or 0's if desired) for all pairs without fingerprints.
- Actual score calculations where optimized using ``numba`` (which turns out to be notably faster than the scipy implementation).

To get the molecular similarity one would hence run something like:
```python
from matchms.filtering import add_fingerprint
from matchms.similarity import FingerprintSimilarityParallel

spectrums = [add_fingerprint(s) for s in spectrums]
similarity_measure = FingerprintSimilarityParallel(similarity_measure="jaccard")
score_matrix = FingerprintSimilarityParallel(spectrums, spectrums)
```

### CODE REVIEW

**Comment by fdiblen** [comment]
@florian-huber : I have just figured out that the isort issue is related to line wrapping. This is most probably because of inchi lines.

**Comment by cwmeijer** [comment]
The isort test is failing because of the multiple imports on a single line. We don't allow that with the current isort settings. Just run 'isort --recursive .' and it will fix it for you. See below for the changes that it will make.

```
C:\Users\ChristiaanMeijer\Documents\matchms>isort --check-only --diff --recursive .
ERROR: C:\Users\ChristiaanMeijer\Documents\matchms\matchms\similarity\FingerprintSimilarityParallel.py Imports are incorrectly sorted.
--- C:\Users\ChristiaanMeijer\Documents\matchms\matchms\similarity\FingerprintSimilarityParallel.py:before      2020-05-29 14:04:49.774187
+++ C:\Users\ChristiaanMeijer\Documents\matchms\matchms\similarity\FingerprintSimilarityParallel.py:after       2020-05-29 14:09:08.989595
@@ -1,8 +1,10 @@
 import numpy
-
-from matchms.similarity.vector_similarity_functions import (cosine_similarity_matrix,
-                                                            dice_similarity_matrix,
-                                                            jaccard_similarity_matrix)
+from matchms.similarity.vector_similarity_functions import \
+    cosine_similarity_matrix
+from matchms.similarity.vector_similarity_functions import \
+    dice_similarity_matrix
+from matchms.similarity.vector_similarity_functions import \
+    jaccard_similarity_matrix


 class FingerprintSimilarityParallel:
ERROR: C:\Users\ChristiaanMeijer\Documents\matchms\tests\test_utils.py Imports are incorrectly sorted.
--- C:\Users\ChristiaanMeijer\Documents\matchms\tests\test_utils.py:before      2020-05-29 14:04:49.776147
+++ C:\Users\ChristiaanMeijer\Documents\matchms\tests\test_utils.py:after       2020-05-29 14:09:09.109276
@@ -1,8 +1,10 @@
 import numpy
-
-from matchms.utils import (derive_fingerprint_from_inchi,
-                           derive_fingerprint_from_smiles, is_valid_inchi,
-                           is_valid_inchikey, is_valid_smiles, mol_converter)
+from matchms.utils import derive_fingerprint_from_inchi
+from matchms.utils import derive_fingerprint_from_smiles
+from matchms.utils import is_valid_inchi
+from matchms.utils import is_valid_inchikey
+from matchms.utils import is_valid_smiles
+from matchms.utils import mol_converter


 def test_mol_converter_smiles_to_inchi():

C:\Users\ChristiaanMeijer\Documents\matchms>isort --recursive .
Fixing C:\Users\ChristiaanMeijer\Documents\matchms\matchms\similarity\FingerprintSimilarityParallel.py
Fixing C:\Users\ChristiaanMeijer\Documents\matchms\tests\test_utils.py
Skipped 2 files
```

**Comment by florian-huber** [comment]
@cwmeijer with this I know how to fix the isort issue. Thanks!

**Comment by florian-huber** [comment]
Too me, solving the isort complaints was a bit confusing, I guess mostly because the settings in ``setup.cfg`` changed in the past. That let to conflicting suggestions/corrections when running ``isort --recursive``. 
The import type that caused trouble was for instance done by a former ``isort`` run:
```
from matchms.similarity.vector_similarity_functions import (cosine_similarity_matrix,
                                                            dice_similarity_matrix,
                                                            jaccard_similarity_matrix)
```
I now changed it by running isort and got the two changes Christiaan showed above.

**Comment by cwmeijer** [comment]
When I read your last comment, it looks like you managed to fix the issue with isort, but in the latest checks isort is still complaining. It is complaining however about unexpected issues like TypeErrors (https://github.com/matchms/matchms/pull/45/checks?check_run_id=730046063).

Long story short: Was your isort issue above fixed? Do you need more help with this?

**Comment by florian-huber** [comment]
Thanks @cwmeijer , I was simply too optimistic with my last changes :)
 #56 now solved this!

File: matchms/filtering/add_fingerprint.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+from ..typing import SpectrumType
+from ..utils import derive_fingerprint_from_inchi
+from ..utils import derive_fingerprint_from_smiles
+
+
+def add_fingerprint(spectrum_in: SpectrumType, fingerprint_type: str = "daylight",
+                    nbits: int = 2048) -> SpectrumType:
+    """Add molecular finterprint to spectrum.
+
+    If inchi or smiles present in metadata, derive a molecular finterprint and
```
Comment (sverhoeven) [inline matchms/filtering/add_fingerprint.py]:
```suggestion
    If smiles or inchi present in metadata, derive a molecular finterprint and
```

**Review by sverhoeven — APPROVED** [review]

**Pull request merged**